runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/testing/TestDataflowRunner.java

```diff
@@ -174,24 +183,29 @@ DataflowPipelineJob run(Pipeline pipeline, DataflowRunner runner) {
     return runner.apply(transform, input);
   }
 
-  Optional<Boolean> checkForSuccess(DataflowPipelineJob job)
+  /**
+   * Check success status of a dataflow job with given job metrics.
+   *
+   * @return Optional.of(false) if job failed/cancelled or PAssert failed,
+   *         Optional.of(true) if number of success PAssert meet expects,
+   *         or Optional.absent() otherwise.
+   */
+  Optional<Boolean> checkForSuccess(DataflowPipelineJob job, JobMetrics metrics)
       throws IOException {
     State state = job.getState();
     if (state == State.FAILED || state == State.CANCELLED) {
       LOG.info("The pipeline failed");
       return Optional.of(false);
     }
 
-    JobMetrics metrics = options.getDataflowClient().projects().jobs()
-        .getMetrics(job.getProjectId(), job.getJobId()).execute();
-
     if (metrics == null || metrics.getMetrics() == null) {
       LOG.warn("Metrics not present for Dataflow job {}.", job.getJobId());
```

Comment:
comparing to have same return value in multiple places, we prefer to avoid large else block?